src/constraints/joints/distance.rs

```diff
@@ -0,0 +1,194 @@
+//! [`DistanceJoint`] component.
+
+use crate::prelude::*;
+use bevy::prelude::*;
+
+/// A distance joint prevents relative translation of the attached bodies away from its rest length while allowing rotation around all axes.
+///
+/// Distance joints can be useful for things like springs, muscles, and mass-spring networks.
+#[derive(Component, Clone, Copy, Debug, PartialEq)]
+pub struct DistanceJoint {
+    /// First entity constrained by the joint.
+    pub entity1: Entity,
+    /// Second entity constrained by the joint.
+    pub entity2: Entity,
+    /// Attachment point on the first body.
+    pub local_anchor1: Vector,
+    /// Attachment point on the second body.
+    pub local_anchor2: Vector,
+    /// The distance the attached bodies will be kept relative to each other.
+    pub rest_length: Scalar,
+    /// The extents of the allowed relative translation between the attached bodies.
+    pub length_limits: Option<DistanceLimit>,
+    /// Linear damping applied by the joint.
+    pub damping_linear: Scalar,
+    /// Angular damping applied by the joint.
+    pub damping_angular: Scalar,
+    /// Lagrange multiplier for the positional correction.
+    pub lagrange: Scalar,
+    /// The joint's compliance, the inverse of stiffness, has the unit meters / Newton.
+    pub compliance: Scalar,
+    /// The force exerted by the joint.
+    pub force: Vector,
+}
+
+impl XpbdConstraint<2> for DistanceJoint {
+    fn entities(&self) -> [Entity; 2] {
+        [self.entity1, self.entity2]
+    }
+
+    fn clear_lagrange_multipliers(&mut self) {
+        self.lagrange = 0.0;
+    }
+
+    fn solve(&mut self, bodies: [&mut RigidBodyQueryItem; 2], dt: Scalar) {
+        self.force = self.constrain_length(bodies, dt);
+    }
+}
+
+impl Joint for DistanceJoint {
+    fn new(entity1: Entity, entity2: Entity) -> Self {
```

Comment:
I noticed that `Joint` must implement `new` like this, but it would probably be nice to be able to specify the rest length here directly since it's the main point of the distance joint.

This is fine for now, but I can make a follow-up PR that removes `new` from `Joint` to allow joints to have custom constructor methods.

Reply:
I think a distance joint with zero rest length is valid though, so I wouldn't be eager to remove new (). It's nice to have a unified constructor for all joints.

Reply:
It's valid, but most people probably want to specify a non-zero rest length for distance joints, otherwise you could use some other joint like a spherical joint. But I'll leave it as it is for now, I agree that it's nice to have a unified API.